internal/v1/handler_compose_image.go

```diff
@@ -180,6 +190,90 @@ func buildRepositories(arch *distribution.Architecture, imageType ImageTypes) []
 	return repositories
 }
 
+func (h *Handlers) buildRepositorySnapshots(ctx echo.Context, arch *distribution.Architecture, imageType ImageTypes, snapshotDate *string) ([]composer.Repository, error) {
+	date, err := time.Parse(time.RFC3339, *snapshotDate)
+	if err != nil {
+		return nil, echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Snapshot date %s is not in RFC3339 format", *snapshotDate))
+	}
+
+	repoURLs := []string{}
+	for _, r := range arch.Repositories {
+		contains := len(r.ImageTypeTags) == 0
+		for _, it := range r.ImageTypeTags {
+			if it == string(imageType) {
+				contains = true
+				break
+			}
+		}
+		if contains {
+			repoURLs = append(repoURLs, *r.Baseurl)
+		}
+	}
+
+	repoUUIDs := []string{}
+	repoMap := map[string]content_sources.ApiRepositoryResponse{}
+	resp, err := h.server.csClient.GetRepositories(ctx.Request().Context(), repoURLs, false)
+	if err != nil {
+		return nil, err
+	}
+	defer closeBody(ctx, resp.Body)
+
+	var csRepos content_sources.ApiRepositoryCollectionResponse
+	err = json.NewDecoder(resp.Body).Decode(&csRepos)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, repo := range *csRepos.Data {
+		repoUUIDs = append(repoUUIDs, *repo.Uuid)
+		repoMap[*repo.Uuid] = repo
+	}
+
+	snapResp, err := h.server.csClient.GetSnapshotsForDate(ctx.Request().Context(), content_sources.ApiListSnapshotByDateRequest{
+		Date:            common.ToPtr(date.UTC().Format(time.DateOnly)),
+		RepositoryUuids: common.ToPtr(repoUUIDs),
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer closeBody(ctx, snapResp.Body)
+
+	var csSnapshots []content_sources.ApiListSnapshotByDateResponse
+	err = json.NewDecoder(snapResp.Body).Decode(&csSnapshots)
+	if err != nil {
+		return nil, err
+	}
+
+	var repositories []composer.Repository
+	for _, snap := range csSnapshots {
+		repo, ok := repoMap[*snap.RepositoryUuid]
+		if !ok {
+			return repositories, fmt.Errorf("Returned snapshot %v unexpected repository id %v", *snap.Match.Uuid, *snap.RepositoryUuid)
+		}
+
+		composerRepo := composer.Repository{
+			Baseurl: common.ToPtr(h.server.csReposURL.JoinPath(*snap.Match.RepositoryPath).String()),
+			Rhsm:    common.ToPtr(false),
+		}
+
+		composerRepo.Gpgkey = repo.GpgKey
+		if composerRepo.Gpgkey != nil && *composerRepo.Gpgkey != "" {
+			composerRepo.CheckGpg = common.ToPtr(true)
+		}
+		composerRepo.ModuleHotfixes = repo.ModuleHotfixes
+		composerRepo.CheckRepoGpg = repo.MetadataVerification
+
+		repositories = append(repositories, composerRepo)
+	}
+
+	// A sanity check to make sure there's a snapshot for each repo
+	expected := len(buildRepositories(arch, imageType))
+	if len(repositories) != expected {
+		return repositories, fmt.Errorf("No snapshots found for all repositories (found %d, expected %d)", len(repositories), expected)
+	}
+	return repositories, nil
+}
+
```

Comment:
Allright, I really hope the next commit introduces a test for this :-D

I would not mind seeing little bit of debug logging in this function.